### ISSUE TITLE
Can't explicitly validate ControlGroup

### ISSUE THREAD

**Issue by ajtowf** [angular/angular#6170]
It would be nice to have the possibility to explicitly call a **.validate()** method on a ControlGroup.
Starting with an invalid form just because the user still hasn't entered anything isn't always ideal.

In [your forms guide](https://angular.io/docs/ts/latest/guide/forms.html) it's quite convenient to start out with a populated model, hence you don't need to check !field.dirty or !field.touched in the view.

Please see my [video](https://youtu.be/T__VppOCk04) or check out my [ng2_play repo](https://github.com/ajtowf/ng2_play), it's always easier to show the problem.

**Comment by alexpods** [comment]
Maybe [`.updateValueAndValidity()`](https://github.com/angular/angular/blob/master/modules/angular2/src/common/forms/model.ts#L116) is what you're looking for?

**Comment by jlmagee** [comment]
IMHO, it would be easy enough to create an appropriate .validate() method using the discrete form and control properties and call it at the appropriate point in the user interaction if it is undesirable to have dynamic validation (which it may be for some applications).  

A generic .validate() method would require application specific configuration which would probably be more confusing than a little code.

**Comment by alexanderchr** [comment]
This is bothering me too. I've 'solved' it by adding a `forceValidation` flag to my form components which is set to true when the form is submitted. Error styles may then applied with `touched || forceValidation`.

I think a much more sensible default would be to consider empty and untouched fields valid until the form has been validated.

**Comment by DzmitryShylovich** [comment]
isn't `FormGroup.updateValueAndValidity()` enough?

**Comment by DzmitryShylovich** [comment]
@pkozlowski-opensource close?

**Comment by DzmitryShylovich** [comment]
@kara there's `FormGroup.updateValueAndValidity()` so I think it can be closed

**Comment by chaowentan** [comment]
Like @DzmitryShylovich suggested, this limitation can be worked around by using `updateValueAndValidity()`. But beware that `updateValueAndValidity()` doesn't recursively update the values and validity of the descendants; only those of the ancestors:

https://github.com/angular/angular/blob/5e840e1e798f41a5e94cdf691c0a38690e9cfc92/packages/forms/src/model.ts#L424

**Comment by huntsberg** [comment]
+1 - Marking them all as dirty seems... dirty.

**Comment by splincode** [comment]
@ajtowf Is the issue relevant?

**Comment by Gerros** [comment]
#19829  :)
Would be great to have this feature... I always have to manually iterate over nested forms to update validity or mark them as dirty etc.

**Comment by xerofun** [comment]
I would definitely enjoy having the ability to update value and validity through descendants as well.  Took me awhile to figure out why it wasn't.
We are tying our forms to a backing service and decoupling the source of truth from the form and as such run into some issues while trying to provide validation errors back to the forms that may include the invalid fields.  We've had to rely on iterating on subscription everywhere.

**Comment by hisham** [comment]
+1. Just ran into this issue and had to manually do a recursive updateValueAndValidity. Seems like running updateValueAndValidity on a form group should run it not only for that form group but also all it's child controls and groups.

**Comment by LuisAlfonsoPonce** [comment]
this.answersQuestion.updateValueAndValidity({ onlySelf: true, emitEvent: true });   // children
  this.activityForm.updateValueAndValidity({ onlySelf: true, emitEvent: true });       // father

**Comment by mackelito** [comment]
> The main problem here, and I've had this issue with almost all the apps I developed to, is that when the user tries to submit a form and there are validation errors, but the user didn't remembered to fill a required field (and not even touched it), the user expects the fields with an error to be marked as invalid. When the user doesn't see the field with the error, they don't know what to do next and thinks the system had an error.
> 
> That's why I see this method a must. This or a good approach to it. Forms should have this behaviour available in some way.

How have you currently solved this in your apps?

**Comment by iwantwin** [comment]
In Angular 12, if you use a <form> element and make use of submit (and optionally reset_ buttons, you can just style your page according to the .ng-submitted class.

```
// Show red borders when submitted
form.ng-submitted input.ng-invalid {
   border: 1px solid red;
}

// Show errors when form was submitted
form.ng-submitted .custom-form-errors{
   display: block;
}
```

You don't have to mark controls dirty anymore for styling purposes, so the state of dirty/touched etc can actually be realistic now.

**Comment by Guerric-P** [comment]
As a workaround: `(<any>form)._updateTreeValidity({ emitEvent: false });`

**Comment by BojanKogoj** [comment]
Any chance `updateTreeValidity` can be made public?

**Comment by alex-kovalchuk** [comment]
```typescript
for (let inner in this.form.controls) {
   this.form.get(inner)?.updateValueAndValidity();
}
```

**Comment by MillerSvt** [comment]
I can't believe that none of the above have even tried to fix it...

**Comment by mackelito** [comment]
> Can't believe such basic feature takes years to be added in this framework.

Is it really such a common requirement? Can't you just split the form in a logical way instead?
Also if you have a more demanding form then use reactive forms instead of template forms..

**Comment by MartinDevillers** [comment]
> > Can't believe such basic feature takes years to be added in this framework.
> 
> Is it really such a common requirement? Can't you just split the form in a logical way instead? Also if you have a more demanding form then use reactive forms instead of template forms..

This is literally the expected default behavior of *any* form that uses validation. Go to any major websites login page (Facebook; Netflix; Wiki), register a new account, don’t touch the input fields and click the “Submit” button. All the required fields immediately light up with error messages under them. That there is no straightforward way to accomplish this basic UX pattern with this library is an oversight.

**Comment by mackelito** [comment]
> This is literally the expected default behavior of _any_ form that uses validation. Go to any major websites login page (Facebook; Netflix; Wiki), register a new account, don’t touch the input fields and click the “Submit” button. All the required fields immediately light up with error messages under them. That there is no straightforward way to accomplish this basic UX pattern with this library is an oversight.

Now I do agree with the fact that there are rooms for improvement... but to achieve this is all that hard...
I might have misunderstood you @MartinDevillers but is this what you where looking for?
https://stackblitz.com/edit/nn5qsk?file=src%2Fapp%2Fapp.component.ts

**Comment by JeanMeche** [comment]
I agree that this usecase is covered by invoking `this.form.markAllAsTouched()`.

**Comment by hakimio** [comment]
@JeanMeche 
You workaround doesn't work: https://stackblitz.com/edit/angular-k8tplw?file=src%2Fapp%2Fapp.component.ts
See the difference by uncommenting `(this.testForm as any)._updateTreeValidity();` and pressing `Submit`.

**Comment by JeanMeche** [comment]
I see no difference between having `_updateTreeValidity` or not. What am I expected to see?

**Comment by hakimio** [comment]
@JeanMeche calling `this.changeDetector.markForCheck()` or `this.changeDetector.detectChanges()` after `markAllAsTouched()` still don't make any difference for me. What's the fixed version?

**Comment by JeanMeche** [comment]
Could you provide an update-to-date example ? (v17)

**Comment by jlmagee** [comment]
This issue was originally reported against the original Angular in  2015.  Please move this dialog to a more appropriate thread.

**Comment by hakimio** [comment]
@JeanMeche 
Here is an up-to-date example with Angular v17: https://stackblitz.com/edit/angular-6fjyuv?file=src%2Fapp%2Fapp.component.ts

@jlmagee feel free to unsubscribe from the thread if it's bothering you.

**Comment by JeanMeche** [comment]
This looks like an issue with TaigaUI. 
The form control is invalid but this is not reflected in the UI. 

```
    this.testForm.markAllAsTouched();
    console.log(this.testForm.controls.textValue.valid) // false
```

**Comment by hakimio** [comment]
Why is it reflected after calling `_updateTreeValidity()` and what's the non-private replacement for this call?

**Comment by JeanMeche** [comment]
You're right,

Here a material example has a similar issue : https://stackblitz.com/edit/spdpnj?file=src%2Fexample%2Finput-form-example.ts,src%2Fexample%2Finput-form-example.html

**Comment by hakimio** [comment]
@JeanMeche in the Material demo you forgot to add `(click)` listener on the button. So, I agree that it's an issue with "Taiga UI", but still would be interesting to know if there is any real public replacement for `_updateTreeValidity()` since in some cases `markAllAsTouched()` is not enough.

**Comment by JeanMeche** [comment]
Well, then it's is definitly related to how Taiga handles that internally. The fact the submit doesn't do anyting and that markAllForCheck has a delayed effect (like CD related) show that there issues on their side. I would advise to open an issue on them. 

@MartinDevillers I'd like to correct you, 
This is exactly what the `FormGroupDirective` onSubmit does. 

The example I gave above (https://stackblitz.com/edit/spdpnj?file=src%2Fexample%2Finput-form-example.ts,src%2Fexample%2Finput-form-example.html) shows it well. 

I'm adding the needs repro label as we're still missing a real use-case with a reproduction where you need to explicitly update the child tree. (As in the framework doesn't do what it should: automatically updating)
